File: src/tmp/Java/src/com/dbc/algorithm/String/StringToIntegerAtoi.java
package com.dbc.algorithm.String;

import java.util.HashMap;
import java.util.Map;

public class StringToIntegerAtoi {
    private String state = "start";
    private long result = 0;
    private int flag = 1;
    private Map<String, String[]> table = new HashMap<String, String[]>(){{
        put("start", new String[]{"start", "signed", "in_number", "end"});
        put("signed", new String[]{"end", "end", "in_number", "end"});
        put("in_number", new String[]{"end", "end", "in_number", "end"});
        put("end", new String[]{"end", "end", "end", "end"});
    }};

    private int get_col(char c) {
        if (c == ' ') {
            return 0;
        }
        if (c == '+' || c == '-') {
            return 1;
        }
        if (Character.isDigit(c)) {
            return 2;
        }
        return 3;
    }

    public void get(char c) {
        state = table.get(state)[get_col(c)];
        if ("in_number".equals(state)) {
            result = result * 10 + c - '0';
            result = flag == 1 ? Math.min(result, (long) Integer.MAX_VALUE) : Math.min(result, -(long) Integer.MIN_VALUE);
        } else if ("signed".equals(state)) {
            flag = c == '+' ? 1 : -1;
        }
    }

    public int myAtoi(String s) {
        int length = s.length();
        for (int i = 0; i < length; ++i) {
            get(s.charAt(i));
        }
        return (int) (flag * result);
    }
}

File: core/StringToIntegerAtoi/StringToIntegerAtoi.py
def myAtoi(s: str) -> int:
    left, length = 0, len(s)
    while left < length and s[left] == " ":
        left += 1

    if left >= length:
        return 0

    if s[left] in ["-", "+"]:
        right = left + 1
    else:
        right = left

    while right < length and s[right].isdigit():
        right += 1

    if left == right or (left == right - 1 and s[left] in ["-", "+"]):
        return 0
    res = int(s[left:right])
    if res < -2 ** 31:
        return -2 ** 31
    if res > 2 ** 31 - 1:
        return 2 ** 31 - 1
    return int(s[left:right])


